django/test/testcases.py

```diff
@@ -148,6 +148,27 @@ def __call__(self):
         raise AssertionError(self.message)
 
 
+class _CaptureOnCommitCallbacksContext:
+    def __init__(self, *, using=DEFAULT_DB_ALIAS, execute=False):
+        self.using = using
+        self.execute = execute
+        self.callbacks = None
+
+    def __enter__(self):
+        if self.callbacks is not None:
+            raise RuntimeError("Cannot re-enter captureOnCommitCallbacks()")
+        self.start_count = len(connections[self.using].run_on_commit)
+        self.callbacks = []
+        return self.callbacks
+
+    def __exit__(self, exc_type, exc_valuei, exc_traceback):
+        run_on_commit = connections[self.using].run_on_commit[self.start_count:]
+        self.callbacks[:] = [func for sids, func in run_on_commit]
+        if exc_type is None and self.execute:
+            for callback in self.callbacks:
+                callback()
```

Comment:
Maybe I'm missing sth, but can we use `@contextmanager`?
```python
class TestCase(TransactionTestCase):
    ...

    @contextmanager
    def captureOnCommitCallbacks(self, *, using=DEFAULT_DB_ALIAS, execute=False):
        self.callbacks = []
        start_count = len(connections[using].run_on_commit)
        try:
            yield self.callbacks
        finally:
            run_on_commit = connections[using].run_on_commit[start_count:]
            self.callbacks[:] = [func for sids, func in run_on_commit]
            if execute:
                for callback in self.callbacks:
                    callback()
```

Reply:
Yes we can. I think the way this evolved, the design originally wouldn't work with that, but now it can.

I've updated based on your suggestion with some modifications (no `self.callbacks`, `@classmethod`).

The test for prevention of re-entry failed, but this seems to be because `@contextmanager`s seem to already raise an `AttributeError` if you try to re-enter them. So, we don't need our own guard, and I've removed the test:

```
======================================================================
ERROR: test_cannot_reenter (test_utils.tests.CaptureOnCommitCallbacksTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/chainz/Documents/Projects/django/tests/test_utils/tests.py", line 1320, in test_cannot_reenter
    with manager:
  File "/Users/chainz/.pyenv/versions/3.8.3/lib/python3.8/contextlib.py", line 111, in __enter__
    del self.args, self.kwds, self.func
AttributeError: args
```